### PR TITLE
feature/AKD-138,129,130: Added Azure 'VM Backups Enabled', 'VM DailyBackup Retention Period', 'VM Instant Restore Backup Retention Period' plugin and test cases

### PR DESCRIPTION
**Pull request by ali-imran7**

### CODE REVIEW

**Comment by ali-imran7** [comment]
@AkhtarAmir @giorod3 This PR is extension of #661 so pulled those changes here also to get collector calls and helper changes.

File: exports.js [hunk 1]
```diff
@@ -383,6 +383,8 @@ module.exports = {
         'scaleSetMultiAz'               : require(__dirname + '/plugins/azure/virtualmachines/scaleSetMultiAz.js'),
         'scaleSetAutoscaleEnabled'      : require(__dirname + '/plugins/azure/virtualmachines/scaleSetAutoscaleEnabled.js'),
         'vmAvailabilitySetLimit'        : require(__dirname + '/plugins/azure/virtualmachines/vmAvailabilitySetLimit.js'),
+        'dailyBackupRetention'          : require(__dirname + '/plugins/azure/virtualmachines/dailyBackupRetention.js'),
```
Comment (AkhtarAmir) [inline exports.js:386]:
```suggestion
        'vmDailyBackupRetention'          : require(__dirname + '/plugins/azure/virtualmachines/vmDailyBackupRetention.js'),
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    category: 'Virtual Machines',
+    description: 'Ensures that VM daily backup retention policy is configured to retain backups for the desired number of days.',
+    more_info: 'Azure Backup provides independent and isolated backups to guard against unintended destruction of the data on your VMs. These backups should be retained for a specific amount of time to recover destroyed VM.',
+    recommended_action: 'Configure virtual machine daily backup retention policy to retain backups for desired number of days',
+    link: 'https://docs.microsoft.com/en-us/azure/backup/backup-azure-vms-introduction',
+    apis: ['virtualMachines:listAll', 'recoveryServiceVaults:listBySubscriptionId', 'backupProtectedItems:listByVault', 'backupPolicies:listByVault'],
+    settings: {
+        vm_daily_backup_retention_period: {
+            name: 'VM Daily Backup Retention Period',
+            description: 'Number of days that a VM backup will be retained until it is permanently deleted',
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:15]:
```suggestion
            description: 'Number of days that a VM backup will be retained until it is permanently deleted (7 or above)',
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+            const vmPoliciesMap = new Map();
+            const backupPoliciesMap = new Map();
+
+            for (const vault of recoveryVaults.data) {
+                const backupProtectedItems = helpers.addSource(cache, source,
+                    ['backupProtectedItems', 'listByVault', location, vault.id]);
+
+                if (!backupProtectedItems || backupProtectedItems.err || !backupProtectedItems.data) {
+                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(recoveryVaults), location);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:67]:
```suggestion
                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(backupProtectedItems), location);
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                    return rcb();
+                }
+
+                const backupPolicies = helpers.addSource(cache, source,
+                    ['backupPolicies', 'listByVault', location, vault.id]);
+
+                if (!backupPolicies || backupPolicies.err || !backupPolicies.data) {
+                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(recoveryVaults), location);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:75]:
```suggestion
                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(backupPolicies), location);
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                    return rcb();
+                }
+
+                for (const bpItem of backupProtectedItems.data) {
+                    if (vmPoliciesMap.get(bpItem.virtualMachineId.toLowerCase())) {
+                        vmPoliciesMap.get(bpItem.virtualMachineId.toLowerCase()).push(bpItem.policyId);
+                    } else {
+                        vmPoliciesMap.set(bpItem.virtualMachineId.toLowerCase(), []);
+                        vmPoliciesMap.get(bpItem.virtualMachineId.toLowerCase()).push(bpItem.policyId);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:84]: Can we do this?
```suggestion
                        vmPoliciesMap.set(bpItem.virtualMachineId.toLowerCase(), [bpItem.policyId]);
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            async.each(virtualMachines.data, function(virtualMachine, scb) {
+                const vmPolicies = vmPoliciesMap.get(virtualMachine.id.toLowerCase());
+                if (vmPolicies && vmPolicies.length) {
+                    let retentionDays = 0;
+                    for (const vmPolicy of vmPolicies) {
+                        const backupPolicy = backupPoliciesMap.get(vmPolicy);
+
+                        if (backupPolicy && backupPolicy.retentionPolicy && backupPolicy.retentionPolicy.dailySchedule &&
+                            backupPolicy.retentionPolicy.dailySchedule.retentionDuration && backupPolicy.retentionPolicy.dailySchedule.retentionDuration.count &&
+                            backupPolicy.retentionPolicy.dailySchedule.retentionDuration && backupPolicy.retentionPolicy.dailySchedule.retentionDuration.count > retentionDays) {
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:102]:
```suggestion
                            backupPolicy.retentionPolicy.dailySchedule.retentionDuration && backupPolicy.retentionPolicy.dailySchedule.retentionDuration.count && backupPolicy.retentionPolicy.dailySchedule.retentionDuration.count > retentionDays) {
```

File: plugins/azure/virtualmachines/instantRestoreRetention.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+            const vmPoliciesMap = new Map();
+            const backupPoliciesMap = new Map();
+
+            for (const vault of recoveryVaults.data) {
+                const backupProtectedItems = helpers.addSource(cache, source,
+                    ['backupProtectedItems', 'listByVault', location, vault.id]);
+
+                if (!backupProtectedItems || backupProtectedItems.err || !backupProtectedItems.data) {
+                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(recoveryVaults), location);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/instantRestoreRetention.js:67]: 🥇 
```suggestion
                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(backupProtectedItems), location);
```

File: plugins/azure/virtualmachines/instantRestoreRetention.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                    return rcb();
+                }
+
+                const backupPolicies = helpers.addSource(cache, source,
+                    ['backupPolicies', 'listByVault', location, vault.id]);
+
+                if (!backupPolicies || backupPolicies.err || !backupPolicies.data) {
+                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(recoveryVaults), location);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/instantRestoreRetention.js:75]:
```suggestion
                    helpers.addResult(results, 3, 'Unable to query for backup retention policies : ' + helpers.addError(backupPolicies), location);
```

File: plugins/azure/virtualmachines/instantRestoreRetention.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+                const vmPolicies = vmPoliciesMap.get(virtualMachine.id.toLowerCase());
+                if (vmPolicies && vmPolicies.length) {
+                    let retentionDays = 0;
+                    for (const vmPolicy of vmPolicies) {
+                        const backupPolicy = backupPoliciesMap.get(vmPolicy);
+
+                        if (backupPolicy && backupPolicy.instantRpRetentionRangeInDays) {
+                            retentionDays = backupPolicy.instantRpRetentionRangeInDays;
+                        } else {
+                            helpers.addResult(results, 0, 'No instant restore backup is configured', location, virtualMachine.id);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/instantRestoreRetention.js:103]: Never do this in a for loop

File: plugins/azure/virtualmachines/vmBackupsEnabled.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            }
+
+            const vmPoliciesMap = new Map();
+
+            for (const vault of recoveryVaults.data) {
+                const backupProtectedItems = helpers.addSource(cache, source,
+                    ['backupProtectedItems', 'listByVault', location, vault.id]);
+
+                if (!backupProtectedItems || backupProtectedItems.err || !backupProtectedItems.data) {
+                    helpers.addResult(results, 3, 'Unable to query for backups : ' + helpers.addError(recoveryVaults), location);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmBackupsEnabled.js:54]:
```suggestion
                    helpers.addResult(results, 3, 'Unable to query for backups : ' + helpers.addError(backupProtectedItems), location);
```

File: plugins/azure/virtualmachines/vmBackupsEnabled.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+                    }
+                }
+            }
+
+            for (const virtualMachine of virtualMachines.data) {
+                const vmPolicies = vmPoliciesMap.get(virtualMachine.id.toLowerCase());
+
+                let vmBackupsEnabled = false;
+                if (vmPolicies && vmPolicies.length) {
+                    vmBackupsEnabled = vmPolicies.some(policy => (policy && policy !== ''));
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmBackupsEnabled.js:73]:
```suggestion
                    vmBackupsEnabled = vmPolicies.some(policy => (policy && policy.length));
```

File: plugins/azure/virtualmachines/vmBackupsEnabled.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                }
+
+                if (vmBackupsEnabled) {
+                    helpers.addResult(results, 0, 'Azure virtual machine has backups enabled', location, virtualMachine.id);
+                } else {
+                    helpers.addResult(results, 2, 'Azure virtual machine does not have backups enabled', location, virtualMachine.id);
+                }
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmBackupsEnabled.js:80]: Move this outside of for loop

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/azure/virtualmachines/instantRestoreRetention.js [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                    for (const vmPolicy of vmPolicies) {
+                        const backupPolicy = backupPoliciesMap.get(vmPolicy);
+
+                        if (backupPolicy && backupPolicy.instantRpRetentionRangeInDays) {
+                            retentionDays = backupPolicy.instantRpRetentionRangeInDays;
+                        }
+                    }
+
+                    if (retentionDays === 0) {
+                        helpers.addResult(results, 0, 'No instant restore backup is configured', location, virtualMachine.id);
```
Comment (giorod3) [inline plugins/azure/virtualmachines/instantRestoreRetention.js:106]:
```suggestion
                        helpers.addResult(results, 2, 'No instant restore backup is configured', location, virtualMachine.id);
```

File: plugins/azure/virtualmachines/instantRestoreRetention.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+                    for (const vmPolicy of vmPolicies) {
+                        const backupPolicy = backupPoliciesMap.get(vmPolicy);
+
+                        if (backupPolicy && backupPolicy.instantRpRetentionRangeInDays) {
+                            retentionDays = backupPolicy.instantRpRetentionRangeInDays;
+                        }
+                    }
+
+                    if (retentionDays === 0) {
+                        helpers.addResult(results, 2, 'No instant restore backup is configured', location, virtualMachine.id);
```
Comment (giorod3) [inline plugins/azure/virtualmachines/instantRestoreRetention.js:106]: changed to 2

**Pull request merged**